python/ray/util/multiprocessing/pool.py

```diff
@@ -460,11 +589,46 @@ def apply_async(self,
         """
 
         self._check_running()
+        func = self._convert_to_ray_batched_calls_if_needed(func)
         object_ref = self._run_batch(self._random_actor_index(), func,
                                      [(args, kwargs)])
         return AsyncResult(
             [object_ref], callback, error_callback, single_result=True)
 
+    def _convert_to_ray_batched_calls_if_needed(self,
+                                                func: Callable) -> Callable:
+        """Convert joblib's BatchedCalls to RayBatchedCalls for ObjectRef caching.
+
+        This converts joblib's BatchedCalls callable, which is a collection of
+        functions with their args and kwargs to be ran sequentially in an
+        Actor, to a RayBatchedCalls callable, which provides identical
+        functionality in addition to a method which ensures that common
+        args and kwargs are put into the object store just once, saving time
+        and memory. That method is then ran.
+
+        If func is not a BatchedCalls instance, it is returned without changes.
+
+        The ObjectRefs are cached inside two registries (_registry and
+        _registry_hashable), which are common for the entire Pool and are
+        cleaned on close."""
+        if RayBatchedCalls is None:
+            return func
+        org_func = func
+        # SafeFunction is a Python 2 leftover and can be
+        # safely removed.
+        if isinstance(func, SafeFunction):
+            func = func.func
+        if isinstance(func, BatchedCalls):
+            func = RayBatchedCalls(func.items, (func._backend, func._n_jobs),
+                                   func._reducer_callback, func._pickle_cache)
+            # go through all the items and replace args and kwargs with
+            # ObjectRefs, caching them in registries
+            func.put_items_in_object_store(self._registry,
+                                           self._registry_hashable)
+        else:
+            func = org_func
```

Comment:
Would we want a `SafeFunction` to exist here?

Reply:
It was added as a Python 2 workaround, and doesn't seem necessary anymore:
```python
class SafeFunction(object):
    """Wrapper that handles the serialization of exception tracebacks.

    TODO python2_drop: check whether SafeFunction is still needed since we
    dropped support for Python 2. If not needed anymore it should be
    deprecated.

    If an exception is triggered when calling the inner function, a copy of
    the full traceback is captured to make it possible to serialize
    it so that it can be rendered in a different Python process.

    """
```

Therefore, I don't think there is any point to using it. It shouldn't impact any thing one way or another.

Reply:
Ahh okay, Seems fine then :)